library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+ * for KW is 2 to 2^54-1 semiblocks
+ */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
+                ||( in_len > 0x1FFFFFFFFFFFFF8 )
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP800-38F Table 1, plaintext length size limits
+         * for KWP is 1 to 2^32-1 octets
+         */
+        if( ( in_len < 1 )
+#if SIZE_MAX > UINT_MAX
+           || (in_len > 0xFFFFFFFF )
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( (in_len % 8) != 0 )
+        {
+            padlen = ( 8 - ( in_len % 8 ) );
+        }
+
+        memcpy( output, &ICV2, KW_SEMIBLOCK_LENGTH / 2 );
+        PUT_UINT32_BE( ( in_len & 0xffffffff ), output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        memcpy( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t), input, in_len );
+        memset( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t) + in_len, 0, padlen );
+    }
+    semiBlocks = ( ( in_len + padlen ) / KW_SEMIBLOCK_LENGTH ) + 1;
+
+    s = 6 * ( semiBlocks -1 );
+
+    if( ( mode == MBEDTLS_KW_MODE_KWP )
+            && ( in_len <= KW_SEMIBLOCK_LENGTH ) )
+    {
+        memcpy( inBuff, output, 16 );
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx, inBuff, 16, output, &olen);
+        if( ret != 0 )
+            goto cleanup;
+
+        *out_len = olen;
+    }
+    else
+    /*
+     * Do the wrapping function W
+     */
+    {
+        if( semiBlocks < 3 )
+            return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+
+        /* Calculate intermediate values */
+        for( t = 1; t <= s; t++ )
+        {
+            memcpy( inBuff , A, KW_SEMIBLOCK_LENGTH );
+            memcpy( inBuff + KW_SEMIBLOCK_LENGTH, R2, KW_SEMIBLOCK_LENGTH );
+
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx, inBuff, 16, outBuff, &olen);
+            if( ret != 0 )
+                goto cleanup;
+
+            memcpy( A, outBuff, KW_SEMIBLOCK_LENGTH );
+            for( i = 0; i < sizeof(t); i++ )
+            {
+                A[i] ^= ((unsigned char*)&t)[sizeof(t)-1-i];
+            }
+
+            /* shift output semiBlocks  */
+            for( i = 1; i < semiBlocks - 1; i++ )
+                memcpy( output + ( i * KW_SEMIBLOCK_LENGTH ), output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH ), KW_SEMIBLOCK_LENGTH );
+
+            /* Set the last semi block as LSB64 of outBuff*/
+            memcpy( output + ( ( semiBlocks - 1 ) * KW_SEMIBLOCK_LENGTH ), outBuff + KW_SEMIBLOCK_LENGTH, KW_SEMIBLOCK_LENGTH );
+        }
+
+        *out_len = semiBlocks * KW_SEMIBLOCK_LENGTH;
+    }
+cleanup:
+
+    if( ret != 0)
+        mbedtls_platform_zeroize( output, semiBlocks * KW_SEMIBLOCK_LENGTH );
```

Comment:
Do we also need to zeroize `inBuff` and `outBuff`? We've made a copy of potentially sensitive data on this function's stack, and even if the caller chooses to zeroize their copy of it, they don't have a clean way of zeroing our copy.